In [1]:
pip install scikit-learn

In [2]:
pip install datasets

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 493.7/493.7 kB 7.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.3/115.3 kB 7.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 9.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.0/302.0 kB 14.1 MB/s eta 0:00:00


In [3]:
from datasets import load_dataset

imdb_dataset = load_dataset("imdb")['train']

Generating train split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating unsupervised split:   0%|          | 0/50000 [00:00<?, ? examples/s]

In [4]:
train_data = []
train_data_labels = []
for item in imdb_dataset:
  train_data.append(item['text'])
  train_data_labels.append(item['label'])

In [5]:
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer(analyzer='word',max_features=1000,lowercase=True,stop_words='english',ngram_range=(1,2))
features = vectorizer.fit_transform(train_data).toarray()

In [6]:
from sklearn.model_selection import train_test_split
X_train, X_val, y_train, y_val = train_test_split(features,train_data_labels,train_size=0.75,random_state=123)

In [7]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
model_nb = MultinomialNB()
model_dt = DecisionTreeClassifier()
model_rf = RandomForestClassifier()

In [8]:
model_nb = model_nb.fit(X=X_train,y=y_train)
model_dt = model_dt.fit(X=X_train,y=y_train)
model_rf = model_rf.fit(X=X_train,y=y_train)

In [9]:
y_pred_nb = model_nb.predict(X_val)
y_pred_dt = model_dt.predict(X_val)
y_pred_rf = model_rf.predict(X_val)

In [10]:
from sklearn.metrics import accuracy_score, confusion_matrix
print("Naive Bayes", accuracy_score(y_val,y_pred_nb))
print(confusion_matrix(y_val,y_pred_nb))
print()
print("Decision Tree", accuracy_score(y_val,y_pred_dt))
print(confusion_matrix(y_val,y_pred_dt))
print()
print("Random Forest", accuracy_score(y_val,y_pred_rf))
print(confusion_matrix(y_val,y_pred_rf))

Naive Bayes 0.81728
[[2463  622]
 [ 520 2645]]

Decision Tree 0.69952
[[2184  901]
 [ 977 2188]]

Random Forest 0.81952
[[2537  548]
 [ 580 2585]]


Now with the difficult test set

In [11]:
test_set = [
    "Oh, what a delightful experience! I can't contain my excitement, don't know what I enjoyed more, the rude service or the shocking food.",
    "The worst best restaurant I've ever been to. Terrible food and great service.",
    "Compared to a prison cafeteria, this restaurant is amazing! A true culinary wonderland.",
    "If you're into food that defies the laws of physics, this is the place to be.",
    "I can't say I didn't not dislike the food. It's not terrible, maybe.",
    "Not bad, but not great either. It's not like I'll never visit again, but there's no reason to rush back.",
    "Wow, just wow! This place is an absolute dream, if your dream is a nightmare.",
    "The chef here must have a Michelin-starred grandmother. The food is out of this world, literally.",
    "The ambience, if you can even call it that, adds a certain... charm to the dining experience.",
    "I'll give them points for trying. It's definitely a unique dining experience."
]

In [12]:
labels = [0, 0, 0, 0, 0, 1, 1, 1, 1, 1]

In [14]:
vectorizer = CountVectorizer(analyzer='word',max_features=1000,lowercase=True,stop_words='english',ngram_range=(1,2))
features = vectorizer.fit_transform(test_set).toarray()

X_train, X_val, y_train, y_val = train_test_split(features,labels,train_size=0.75,random_state=123)

In [15]:
model_nb = model_nb.fit(X=X_train,y=y_train)
model_dt = model_dt.fit(X=X_train,y=y_train)
model_rf = model_rf.fit(X=X_train,y=y_train)

In [16]:
y_pred_nb = model_nb.predict(X_val)
y_pred_dt = model_dt.predict(X_val)
y_pred_rf = model_rf.predict(X_val)

In [17]:
from sklearn.metrics import accuracy_score, confusion_matrix
print("Naive Bayes", accuracy_score(y_val,y_pred_nb))
print(confusion_matrix(y_val,y_pred_nb))
print()
print("Decision Tree", accuracy_score(y_val,y_pred_dt))
print(confusion_matrix(y_val,y_pred_dt))
print()
print("Random Forest", accuracy_score(y_val,y_pred_rf))
print(confusion_matrix(y_val,y_pred_rf))

Naive Bayes 0.6666666666666666
[[2 0]
 [1 0]]

Decision Tree 0.3333333333333333
[[0 2]
 [0 1]]

Random Forest 0.3333333333333333
[[0 2]
 [0 1]]


Naive Bayes has the highest accuracy among the three models (0.6667). It correctly classifies two instances and misclassifies one instance.

Decision Tree and Random Forest have lower accuracies (0.3333), with Decision Tree misclassifying two instances and Random Forest misclassifying two instances as well.

In this scenario, the ensemble's accuracy matches that of the Decision Tree and Random Forest, but it is lower than the accuracy of the Naive Bayes model. This suggests that the ensemble does not perform better than the best individual model, which is Naive Bayes in terms of overall accuracy.